Init coding style of angular (#29)
C0D1UM/technical-standard: angular/coding-styles/README.md

# Angular Style Guide (Optional)

<br>
<br>

**this page is in draft mode. it is optional to follow guidelines on this page.**

<br>
<br>

## Style Guide

### Base Style Guide

> The main programing language in Angular framework included with Typescript, HTML, SCSS. We follow

Reviewer: ```suggestion
The main programing language in Angular framework included TypeScript, HTML, SCSS. We follow
```

> - [Google style guide for HTML/CSS](https://google.github.io/styleguide/htmlcssguide.html), [Typescript](https://google.github.io/styleguide/tsguide.html)

Reviewer: ```suggestion
- [Google style guide for HTML/CSS](https://google.github.io/styleguide/htmlcssguide.html), [TypeScript](https://google.github.io/styleguide/tsguide.html)
```

- [Kitty Giraudel style guide for SCSS](https://sass-guidelin.es/)
- [Angular Official for Angular](https://angular.io/guide/styleguide).

The rest of the document describes additions and clarifications below.

### Typescript

#### Indentation

> The indentation in typescript we use 2 spaces for every new line. we don't tabs or mix tabs.

Reviewer: ```suggestion
For the indentation in TypeScript, we use **2 spaces** for every new line. we don't use tabs or spaces mixed with tabs.
```

What do you think krub?

Reviewer 2: Changes:
... for every new line. Do not use tabs or spaces mixed with tabs.

#### Return Type

We recommend to assign ***return type*** of method for precision and simple to read the surface of them. Some lint configuration as eslint will have warning with highlight.

```typescript

> /** without return type */

Reviewer 2: Let's make it mandatory for return type, so this should be `/** No */`

isAdmin(role: string) {
  return role === 'admin';
}

> /** recommend */

Reviewer 2: And this should be `/** Yes */`

isAdmin(role: string): boolean {
  return role === 'admin';
}
```

#### Trailing Commas

> If the an array or object literal has written in many line. The last field should be ended with comma.

Reviewer 2: ```suggestion
If an array or object literal is written in many lines, the last object should be ended with comma.
```

```typescript
/** Yes */
const obj = {
  a: 'value1',
  b: 'value1',
  c: 'value1',
}

/** No */
const obj = {
  a: 'value1',
  b: 'value1',
  c: 'value1'
}
```

### HTML

#### Indentation

> The indentation in HTML, we use 2 spaces for inside element. we don't tabs or mix tabs.

Reviewer: ```suggestion
For the indentation in HTML, we use **2 spaces** for elements. we don't use tabs or spaces mixed with tabs.
```

Reviewer 2: Changes:
... for elements. Do not use tabs or spaces mixed with tabs.

```html
<div id="div-1">
  <div id="div-2">
    <div id="div-3">
    </div>
  </div>
</div>
```

#### Line-Wrapping (Optional)

> For wrapping the attribute of element, In base styles there is no recommendation. You may consider by readability. However, we suggest to wrap them with 4 spaces for the attribute indentation.

Reviewer: I think we can give them a constraint krubb.

```suggestion
For wrapping the attribute of elements, we wrap them with 4 spaces for the attribute indentation.
```

```html
<!-- not recommend - wrapping somewhere long line -->
<app-example *ngIf="isOpen" class="text-red" id="example-1" [dropdown]="dropdown" 
  (click)="onClick()" [(data)]="model">
  <button class="btn btn-primary">Hello World!</button>
</app-example>

<!-- Recommend -->
<app-example *ngIf="isOpen"
    class="text-red"
    id="example-1"
    [dropdown]="dropdown"
    (click)="onClick()"
    [(data)]="model">
  <button class="btn btn-primary">Hello World!</button>
</app-example>
```

## Linter

> You should use [`eslint`](https://eslint.org/) instead of [`tslint`](https://palantir.github.io/tslint/) that is deprecated. The `eslint` is available to plug-ins which we suggest to use [`Angular ESLint`](https://github.com/angular-eslint/angular-eslint#migrating-an-angular-cli-project-from-codelyzer-and-tslint).

Reviewer: ```suggestion
You should use [`eslint`](https://eslint.org/) instead of [`tslint`](https://palantir.github.io/tslint/) which is deprecated. `eslint` is available as a plug-in and we suggest you to use [`Angular ESLint`](https://github.com/angular-eslint/angular-eslint#migrating-an-angular-cli-project-from-codelyzer-and-tslint).
```

> After you install `eslint` with Angular plug-in. You may have to  valid some configuration following the style guild.

Reviewer: ```suggestion
After you install `eslint` with Angular plug-in. You may have to valid some configuration following the style guide.
```

## Formatter

> ### For Visual Studio Code (vscode)

Reviewer: ```suggestion
### For Visual Studio Code (VS Code)
```

> For the configuration in vscode, go to Setting -> Extensions and choose the extension file that will configure.

Reviewer: ```suggestion
For the configuration in VS Code, go to `Setting -> Extensions` and choose the extension that we are going to configure.
```

> > NOTE: By default, we can press short key to enter setting UI by `CMD` + `,` (MacOS).

Reviewer: ```suggestion
> NOTE: By default, we can press a short key to enter setting UI by `CMD` + `,` (macOS).
```

- #### HTML

>   These is the configuration that consistent with style guild.

Reviewer: ```suggestion
  These are configurations that consistent with the style guide.
```

  | Title | Value |
  |--------|----------------|
  | `Format: Wrap Attributes` |  `force`  |
  | `Format: Wrap Attributes Indent Size` | `4` |

- #### Typescript

>   For typescript, we just set `TypeScript > Preferences: Quote Style` to `single`.

Reviewer: ```suggestion
  For TypeScript, we just set `TypeScript > Preferences: Quote Style` to `single`.
```

>   We suggest to use `Prettier` extension to enhance formatting.

Reviewer: ```suggestion
  We recommend you to use `Prettier` extension to enhance formatting.
```

>   If you use `Prettier`. There is the configuration following that.

Reviewer: ```suggestion
  If you use `Prettier`, these are configurations that you need to follow.
```

  - `Prettier: Arrow Parent` = `always`
  - `HTML Whitespace Sensitivity` = `strict`

> ### For WebStorm

Reviewer: I'd like to add PyCharm as well krubb.

```suggestion
### For WebStorm/PyCharm
```

> For WebStorm, go to Preference. In sidebar, choose Editor -> Code Style and select programing language to configure.

Reviewer: ```suggestion
For WebStorm/PyCharm, go to Preference. In the sidebar, choose `Editor -> Code Style` and select the programing language to configure.
```

- #### HTML

>   When enter to HTML configuration page. You can set configuration following below.

Reviewer: ```suggestion
  When entering to HTML configuration page. You can set the configuration following below.
```

  - In _Tabs and Indents_ tab, set `indent` = `4`.
  - In _Other_ tab, set `Wrap attributes` = `Wrap always`.

> - #### Typescript

Reviewer: ```suggestion
- #### TypeScript
```

>   In _Punctuation_ tab, you will see three line. set them like below.

Reviewer: ```suggestion
  In _Punctuation_ tab, you will see three lines. Set them like below.
```
